source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -355,6 +313,74 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
+  // is the same thing that we do in STREAMING mode.
+  // 2) If end of stream is reached before the queue reaches the buffer limit, we
+  // send the buffered data to the server and essentially behave as if we are in
+  // buffered mode.
+  // 3) If instead the buffer limit is reached before end of stream, then we also
+  // send the buffered data to the server, and raise the watermark to prevent Envoy
+  // from running out of memory while we wait.
+  // 4) It is possible that Envoy will keep sending us data even in that case, so
+  // we must continue to queue data and prepare to re-inject it later.
+  Http::FilterDataStatus result;
+  if (state.partialBodyProcessed()) {
+    // We already sent and received the buffer, so everything else just falls through.
+    ENVOY_LOG(trace, "Partial buffer limit reached");
+    result = FilterDataStatus::Continue;
+  } else if (state.callbackState() == ProcessorState::CallbackState::BufferedPartialBodyCallback) {
+    // More data came in while we were waiting for a callback result. We need
+    // to queue it and deliver it later in case the callback changes the data.
+    state.enqueueStreamingChunk(data, false, false);
+    ENVOY_LOG(trace, "Call in progress for partial mode");
+    state.setPaused(true);
+    result = FilterDataStatus::StopIterationNoBuffer;
+  } else if (end_stream || state.queueOverHighLimit()) {
+    switch (openStream()) {
+    case StreamOpenState::Error:
+      return {false, FilterDataStatus::StopIterationNoBuffer};
+    case StreamOpenState::IgnoreError:
+      return {false, FilterDataStatus::Continue};
+    case StreamOpenState::Ok:
+      // Fall through
+      break;
+    }
+    state.enqueueStreamingChunk(data, false, false);
+    // Put all buffered data so far into one big buffer
+    const auto& all_data = state.consolidateStreamedChunks(true);
+    ENVOY_LOG(debug, "Sending {} bytes of data in buffered partial mode. end_stream = {}",
+              all_data.data.length(), end_stream);
+    sendBodyChunk(state, all_data.data, ProcessorState::CallbackState::BufferedPartialBodyCallback,
+                  end_stream);
+    result = FilterDataStatus::StopIterationNoBuffer;
+    state.setPaused(true);
+  } else {
+    // Keep on running and buffering
+    state.enqueueStreamingChunk(data, false, false);
+
+    if (Runtime::runtimeFeatureEnabled(Runtime::defer_processing_backedup_streams) &&
+        state.queueOverHighLimit()) {
+      // We just transitioned to queue over high limit, which will then cause
+      // the receiving codec to buffer the data rather than send it to us.
+      // Prior, we'd rely on eager processing to kick off sending the data to
+      // the external processor, instead we kick it off here.
+      // The recursive call here should only be one deep as we should
+      // end up in the branch where we send data in buffered partial mode.
+      Buffer::OwnedImpl empty_buffer{};
+      return processBufferedPartial(state, empty_buffer, false);
```

Comment:
Will this call essentially end up in the `else if (end_stream || state.queueOverHighLimit())` part of the code above?
If that's the case I would suggest extracting that to another function and calling it directly (trying to avoid future changes to the beginning of this function that will change course).

Reply:
Good idea, will extract that instead and invoke it directly from both places.